testsuite/classlibrary/TestCoreUGens.sc

```diff
@@ -4,433 +4,501 @@ UnitTest.gui
+		o = OSCresponderNode(s.addr, '/tr', {|time, resp, msg|
+			switch(msg[2],
+				5453, {testNaN = msg[3] <= 0.0 or:{msg[3] >= 1.0}}		);
+		});
+		o.add;
+		{Line.kr(1, 0, 1, 1, 0, 2); SendTrig.kr(Impulse.kr(10, 0.5), 5453, LFTri.ar(Duty.ar(0.1, 0, Dseq(#[100], 1))))}.play;
+		1.5.wait;
+		s.sync;
+		this.assert(testNaN.not, "Duty+LFTri should not output NaN");
+		o.remove;
+	}
+
+	test_pitchtrackers {
+		var tests = Dictionary[
+			"ZCR.ar() tracking a SinOsc"
 			-> { var freq = XLine.kr(100, 1000, 10);
 				var son = SinOsc.ar(freq);
 				var val = A2K.kr(ZeroCrossing.ar(son));
 				var dev = (freq-val).abs * XLine.kr(0.0001, 1, 0.1);
 				Out.ar(0, (son * 0.1).dup);
 				dev},
-		"Pitch.kr() tracking a Saw"
-				-> { var freq = XLine.kr(100, 1000, 10);
+			"Pitch.kr() tracking a Saw"
+			-> { var freq = XLine.kr(100, 1000, 10);
 				var son = Saw.ar(freq);
 				var val = Pitch.kr(son).at(0);
 				var dev = (freq-val).abs * XLine.kr(0.0001, 1, 0.1);
 				Out.ar(0, (son * 0.1).dup);
 				dev * 0.1 /* rescaled cos Pitch more variable than ZCR */ },
 		];
-	var testsIncomplete = tests.size;
-	this.bootServer;
-	tests.keysValuesDo{|text, func|
-		func.loadToFloatArray(10, Server.default, { |data|
-			this.assertArrayFloatEquals(data, 0.0, text, within: 1.0);
-			testsIncomplete = testsIncomplete - 1;
-		});
-		rrand(0.12, 0.35).wait;
-	};
+		var testsIncomplete = tests.size;
+		this.bootServer;
+		tests.keysValuesDo{|text, func|
+			func.loadToFloatArray(10, Server.default, { |data|
+				this.assertArrayFloatEquals(data, 0.0, text, within: 1.0);
+				testsIncomplete = testsIncomplete - 1;
+			});
+			rrand(0.12, 0.35).wait;
+		};
+
+		// Wait for async tests
+		this.wait{testsIncomplete==0};
+	} // test_pitchtrackers
+
+	test_out_ugens {
+		var testAudioRate, testControlRate;
+
+		testAudioRate = {
+			var failing = [nil, []];
+			var working = [0, [0], [0, 0, 0]];
+			var stubs = [
+				{ |args| LocalOut.kr(*args) },
+				{ |args| Out.kr(0, *args) },
+				{ |args| ReplaceOut.kr(0, *args) },
+				{ |args| XOut.kr(0, 0.5, *args) },
+			];
+
+			var fails = failing.every { |input|
+				stubs.every { |func|
+					var fails = false;
+					try { { func.(input) }.asSynthDef } { |err| fails = err.isException };
+					if(fails.not) { "This is a case that should have failed:\n%\nOn this input:\n%\n".postf(func.cs, input) };
+					fails
+				}
+			};
+			var works = working.every { |input|
+				stubs.every { |func|
+					var works = true;
+					try { { func.(input) }.asSynthDef } { |err| works = err.isException.not };
+					if(works.not) { "This is a case that should not have failed:\n%\nOn this input:\n%\n".postf(func.cs, input) };
+					works
+				}
+			};
+
+			fails and: works
+		};
+
+		testControlRate = {
+			var failing = [nil, []];
+			var working = [{DC.ar(0)}, {DC.ar([0, 0])}];
+			var stubs = [
+				{ |args| LocalOut.ar(*args) },
+				{ |args| Out.ar(0, *args) },
+				{ |args| ReplaceOut.ar(0, *args) },
+				{ |args| XOut.ar(0, 0.5, *args) },
```

Comment:
Please add OffsetOut to this list (it's only got an `.ar` method). The test still passes for me if I add it.